lib/spack/spack/cmd/diff.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+import operator
+import sys
+
+import llnl.util.tty as tty
+import llnl.util.tty.color as color
+
+import spack.cmd
+import spack.cmd.common.arguments as arguments
+import spack.environment as ev
+import spack.solver.asp as asp
+import spack.util.environment
+import spack.util.spack_json as sjson
+
+description = "compare two specs"
+section = "basic"
+level = "long"
+
+
+def setup_parser(subparser):
+    arguments.add_common_arguments(
+        subparser, ['specs'])
+
+    subparser.add_argument(
+        '--json',
+        action='store_true',
+        default=False,
+        dest='dump_json',
+        help="Dump json output instead of pretty printing."
+    )
+    subparser.add_argument(
+        '--first',
+        action='store_true',
+        default=False,
+        dest='load_first',
+        help="load the first match if multiple packages match the spec"
+    )
+    subparser.add_argument(
+        '-a',
+        dest='attributes',
+        action='append',
+        help="select the attributes to show (defaults to all)"
+    )
+
+
+def bold(string):
+    """
+    Make a header string bold so we can easily see it
+    """
+    return color.colorize("@*{%s}" % string)
+
+
+def compare_specs(a, b, to_string=False, colorful=True):
+    """
+    Generate a comparison, including diffs (for each side) and an intersection.
+
+    We can either print the result to the console, or parse
+    into a json object for the user to save. We return an object that shows
+    the differences, intersection, and names for a pair of specs a and b.
+
+    Arguments:
+        a (spack.spec.Spec): the first spec to compare
+        b (spack.spec.Spec): the second spec to compare
+        a_name (str): the name of spec a
+        b_name (str): the name of spec b
+        to_string (bool): return an object that can be json dumped
+        colorful (bool): do not format the names for the console
+    """
+    # Prepare a solver setup to parse differences
+    setup = asp.SpackSolverSetup()
+
+    a_facts = set(to_tuple(t) for t in setup.spec_clauses(a))
+    b_facts = set(to_tuple(t) for t in setup.spec_clauses(b))
```

Comment:
Ok I see what I was missing here.  `spec_clauses` returns `*_set` functions by default but that is likely not what we want for concrete specs.

There are two arguments to `spec_clauses` we should probably care about here:

```python
def _spec_clauses(self, spec, body=False, transitive=True):
        """Return a list of clauses for a spec mandates are true.

        Arguments:
            spec (spack.spec.Spec): the spec to analyze
            body (bool): if True, generate clauses to be used in rule bodies
                (final values) instead of rule heads (setters).
            transitive (bool): if False, don't generate clauses from
                 dependencies (default True)
        """
```

I think to keep the facts intuitive and to omit `_set`, you should pass `body=True` here.  `_set` is an implementation detail of the solver -- the "real" names that are used in concrete specs are what you get if you pass `body=True`.


I think in another PR we should add something to allow you to do `transitive=False` and only diff roots, but that is another day.

Reply:
Cool I'll try that.

Reply:
oh that looks much better :)